block-storage/src/main/scala/coop/rchain/blockstorage/dag/BlockDagKeyValueStorage.scala

```diff
@@ -147,11 +150,16 @@ final class BlockDagKeyValueStorage[F[_]: Concurrent: Log] private (
                     else ().pure[F]
                 //Block which contains newly bonded validators will not
                 //have those validators in its justification
-                newValidators = bonds(block)
+                bondingValidators = bonds(block)
                   .map(_.validator)
                   .toSet
                   .diff(block.justifications.map(_.validator).toSet)
-                newValidatorsLatestMessages = newValidators.map(v => (v, block.blockHash))
+                bondingValidatorsLMs = bondingValidators.map(v => (v, block.blockHash))
+                // as we can add blocks in backwards (e.g. when reconstructing finalized state) -
+                // to ot overwrite existing LMs, we should have this check
+                newValidatorsLMs <- bondingValidatorsLMs.toList.filterA(
+                                     lm => latestMessagesIndex.contains(lm._1).not
+                                   )
```

Comment:
This seems like another variant of the same check what is done with `seqNum` few lines below. Can we reuse the same logic to detect if latest message is already with highest seq number?

This can be simplified to use extension with only one key. 
`latestMessagesIndex.get(block.sender)`

Also it would be nice if seq number is kept as Option[Long] and used directly in check without introducing `0` as default.

Reply:
I think the overall `insert` method can be refactored, so was not doing any significant change. Should I make bigger refactoring hdy think?

Reply:
I wanted to minimize changes in migration to LMDB but now it makes sense to refactor parts touched with the fix.

It also deserves separate PR to dev branch.

Reply:
https://github.com/rchain/rchain/pull/3123